### PR TITLE
[Merged by Bors] - feat: lemmas about derivatives

### PR DESCRIPTION
**Pull request by fpvandoorn**
* From the Sobolev project

Co-authored-by: Heather Macbeth 25316162+hrmacbeth@users.noreply.github.com

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/Calculus/ContDiff.lean [hunk 1]
```diff
@@ -1183,6 +1183,15 @@ theorem contDiff_pi : ContDiff 𝕜 n Φ ↔ ∀ i, ContDiff 𝕜 n fun x => Φ
   simp only [← contDiffOn_univ, contDiffOn_pi]
 #align cont_diff_pi contDiff_pi
 
+theorem contDiff_update (k : ℕ∞) (x : ∀ i, F' i) (i : ι) : ContDiff 𝕜 k (update x i) := by
```
Comment (eric-wieser) [inline Mathlib/Analysis/Calculus/ContDiff.lean:1186]: Can you add `single` too as a special case?

**Comment by eric-wieser** [comment]
I don't think git recognizes `Co-authored-by:` if you put it after a bullet point.

File: Mathlib/Analysis/Calculus/Deriv/Pi.lean [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+theorem hasDerivAt_update {x : ι → 𝕜} {i : ι} (y : 𝕜) :
+    HasDerivAt (Function.update x i) (Pi.single i (1:𝕜)) y := by
+  convert (hasFDerivAt_update (E := fun _ ↦ 𝕜) y).hasDerivAt
+  ext z j
+  rw [Pi.single, Function.update_apply]
+  split_ifs with h
+  · simp [h]
+  · simp [Function.update_noteq h]
+
+theorem deriv_update {x : ι → 𝕜} {i : ι} (y : 𝕜) :
```
Comment (eric-wieser) [inline Mathlib/Analysis/Calculus/Deriv/Pi.lean:24]: Can you add the single version of these two as well?

File: Mathlib/Analysis/Calculus/FDeriv/Pi.lean [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    dsimp [Function.update]
+    split_ifs with hji
+    · subst hji
+      simp
+    · simp
+  rw [update_eq]
+  convert (hasFDerivAt_const _ _).add (l.hasFDerivAt.comp y (hasFDerivAt_sub_const (x i)))
+  rw [zero_add, ContinuousLinearMap.comp_id]
+
+theorem fderiv_update {x : ∀ i, E i} {i : ι} (y : E i) :
```
Comment (eric-wieser) [inline Mathlib/Analysis/Calculus/FDeriv/Pi.lean:33]: And the `single` version of these two

File: Mathlib/Analysis/Calculus/FDeriv/Pi.lean [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Floris van Doorn, Heather Macbeth
+-/
+import Mathlib.Analysis.Calculus.FDeriv.Add
+
+/-!
+# Derivatives on pi-types.
+-/
+
+noncomputable section
```
Comment (eric-wieser) [inline Mathlib/Analysis/Calculus/FDeriv/Pi.lean:12]: I don't see any `def`s here, so maybe this can be removed.

**Review by eric-wieser — APPROVED** [review]
bors d+

**Comment by fpvandoorn** [comment]
bors merge

**Pull request closed without merging**